### ISSUE TITLE
XmlPeek not working correctly for element nodes

### ISSUE THREAD

**Issue by peter-perot** [cake-build/cake#1537]
### What You Are Seeing?

I have the following XML document:

``` xml
<?xml version="1.0" encoding="utf-8"?>
<Project ToolsVersion="4.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
  <PropertyGroup>
    <WebPublishMethod>FileSystem</WebPublishMethod>
    <LastUsedBuildConfiguration>DeploymentTemplate</LastUsedBuildConfiguration>
    <LastUsedPlatform>Any CPU</LastUsedPlatform>
    <SiteUrlToLaunchAfterPublish />
    <LaunchSiteAfterPublish>True</LaunchSiteAfterPublish>
    <ExcludeApp_Data>False</ExcludeApp_Data>
    <publishUrl>C:\Deployment\DeploymentTemplate\WebApi</publishUrl>
    <DeleteExistingFiles>False</DeleteExistingFiles>
  </PropertyGroup>
</Project>
```

The following code does not yield the `publishUrl` element (instead `null` ist returned and a warning is logged):

``` csharp
var xmlPeekSettings = new XmlPeekSettings
{
  Namespaces = new Dictionary<string, string>{
    { "msbuild", "http://schemas.microsoft.com/developer/msbuild/2003" }
  }
};

var xmlElementContents = XmlPeek(pubXmlFile,
  "/msbuild:Project/msbuild:PropertyGroup/msbuild:publishUrl", xmlPeekSettings);
```

Running the same code using the standard .NET `XPathDocument` works okay.

### What is Expected?

`XmlPeek` should return the contents of the `publishUrl` element, i.e. `C:\Deployment\DeploymentTemplate\WebApi`.

When using `XPathDocument` everything works okay:

``` csharp
using (var stream = new System.IO.FileStream(pubXmlFile.ToString(), FileMode.Open))
{
  var xPathDocument = new System.Xml.XPath.XPathDocument(stream);
  var navigator = xPathDocument.CreateNavigator();

  System.Xml.XmlNamespaceManager ns = null;

  if (navigator.NameTable != null)
  {
    ns = new System.Xml.XmlNamespaceManager(navigator.NameTable);
    ns.AddNamespace("msbuild", "http://schemas.microsoft.com/developer/msbuild/2003");
  }

  var node = navigator.SelectSingleNode(
    "/msbuild:Project/msbuild:PropertyGroup/msbuild:publishUrl", ns);
}
```

Here `node` contains the selected node.

### What version of Cake are you using?

Version 0.17.0

### Are you running on a 32 or 64 bit system?

64bit

### What environment are you running on?  Windows? Linux? Mac?

Windows

### How Did You Get This To Happen? (Steps to Reproduce)

See code above.

**Comment by peter-perot** [comment]
# _[Proposal for a Bugfix]_

After some investigation I found the bug. Have a look at line https://github.com/cake-build/cake/blob/303b1d092864ae119cbcfb9d3430b99cb8d3bd4f/src/Cake.Common/Xml/XmlPeekAliases.cs#L141:

When calling method `SelectSingleNode(...)` on an instance of `XmlDocument`, an `XmlNode` is returned. Then in turn when reading the [`Value` property](https://msdn.microsoft.com/library/system.xml.xmlnode.value(v=vs.110).aspx), `null` is returned for nodes of type `XmlNodeType.Element` (so say the docs from Microsoft).

**This does not happen if you are using an `XPathNavigator`** - yes, this is really weird, but it works.

Here the corrected code:

``` csharp
/// <summary>
/// Gets the value of a target node.
/// </summary>
/// <returns>The value found at the given XPath query (or the first, if multiple eligible nodes are found).</returns>
/// <param name="source">The source xml to transform.</param>
/// <param name="xpath">The xpath of the nodes to set.</param>
/// <param name="settings">Additional settings to tweak Xml Peek behavior.</param>
private static string XmlPeek(XmlReader source, string xpath, XmlPeekSettings settings)
{
    if (source == null)
    {
        throw new ArgumentNullException(nameof(source));
    }

    if (string.IsNullOrWhiteSpace(xpath))
    {
        throw new ArgumentNullException(nameof(xpath));
    }

    if (settings == null)
    {
        throw new ArgumentNullException(nameof(settings));
    }

    var document = new XmlDocument();
    document.PreserveWhitespace = settings.PreserveWhitespace;
    document.Load(source);

    var navigator = document.CreateNavigator();
    XmlNamespaceManager namespaceManager = null;

    if (navigator.NameTable != null)
    {
        namespaceManager = new XmlNamespaceManager(navigator.NameTable);
        foreach (var xmlNamespace in settings.Namespaces)
        {
            namespaceManager.AddNamespace(xmlNamespace.Key /* Prefix */, xmlNamespace.Value /* URI */);
        }
    }

    var node = navigator.SelectSingleNode(xpath, namespaceManager);

    return node?.Value;
}
```

**Comment by alvaromongon** [comment]
I have just updated to Cake version 0.21.1 and it is still happening.

Any estimation about when this will be fixed?

Thanks!

**Comment by peter-perot** [comment]
Currently I have not the time to drop a pull request, but the code I posted above fixes the issue. Maybe some of the Cake guys can take the code snippet and fix that bug. ;-)

**Comment by devlead** [comment]
fixed by #1701
